### ISSUE TITLE
'$exists' does not work with JSON paths

### ISSUE THREAD

**Issue by kerfab** [kapouille/mongoquery#7]
Given the following object:

    o = { 'a' : { 'b' : { 'c' : True } } }

The query:

    q = Query({ 'a.b' : { '$exists' : True } })

Will result in: 

    q.match(o)
    False

But it works when specifying nested objects rather than a path:

    q = Query({'a' : { 'b' : {'$exists' : True}}})
    q.match(o)
    True

This seems to be due to specific handling and conditions within the package on '$exists'. A fix would be most welcomed! Thank you for that useful library.

**Comment by foolswood** [comment]
There are a couple of approaches that I can think of to fix this:
- Canonicalise the representation of the query on the way in.
- Handle the dotted lookup explicitly for exists checking.

I think the first one may end up being simpler as it should be fairly easy to test and reduce the rest of the code's exposure to errors like this.

**Comment by kapouille** [comment]
@foolswood is right, canonicalisation does seem the best solution.

**Comment by kerfab** [comment]
**I propose the following modifications to __init__.py:**

    def _process_condition(self, operator, condition, entry):
        if isinstance(condition, Mapping) and "$exists" in condition:
            if condition["$exists"] != (operator in entry):
                return False
            elif tuple(condition.keys()) == ("$exists",):
                return True
        if isinstance(operator, string_type):
            if operator.startswith("$"):
                try:
                    return getattr(self, "_" + operator[1:])(condition, entry)
                except AttributeError:
                    raise QueryError(
                        "{!r} operator isn't supported".format(operator))
            else:
                extracted_data = self._extract(entry, operator.split("."))
        else:
            if operator not in entry:
                return False
            extracted_data = entry[operator]
        return self._match(condition, extracted_data)

**becomes:**

    def _process_condition(self, operator, condition, entry):
        if isinstance(operator, string_type):
            if operator.startswith("$"):
                try:
                    return getattr(self, "_" + operator[1:])(condition, entry)
                except AttributeError:
                    raise QueryError(
                        "{!r} operator isn't supported".format(operator))
            else:
                extracted_data = self._extract(entry, operator.split("."))
        else:
            if operator not in entry:
                return False
            extracted_data = entry[operator]
        return self._match(condition, extracted_data)

**Deleted:**

    _exists = _noop

**Added:**

    def _exists(self, condition, entry):
        if not isinstance(entry, _Undefined):
            return condition
        return not condition

**Now it works as intended:**

```
o = {'a':{'b':{'c':1}}}

print(Query({ 'a' : { 'b' : { 'c' : { '$exists' : True } } } }).match(o))  # Old way
True

print(Query({ 'a.b.c' : {'$exists' : True } }).match(o)) # New way
True

print(Query({ 'a.b.f' : {'$exists' : True } }).match(o))
False

print(Query({ 'a.b.f' : {'$exists' : False } }).match(o))
True
```

The `$exists` special handling case is now gone and the function `_process_condition()` is back to genericity. Your review @kapouille is most welcomed, I hope this does not make regressions.

**Comment by foolswood** [comment]
Raise a PR for it, run the tests and see :)

**Comment by kapouille** [comment]
If it passes the tests @kerfab , then we're good :) Please submit a PR with all those nice changes in and I'll create a new version.

**Comment by kerfab** [comment]
Status update: I ran the tests with `nosetests` and it failed when the object is looking like [{1:'value'}, {2:...}]

I think a fix is possible, but might indeed require special handling case in `_process_condition()`

I am working on it, but no guarantee I can elegantly handle this. 

Keep the issue open, I will update whenever I succeed or abandon.

**Comment by kapouille** [comment]
Good luck @kerfab , keep us updated!

**Comment by kerfab** [comment]
Status update: I fixed it, and AFAIK it works.

Travis tests fail on one test due to travis itself.

It would be cool if the `nosetests` include some more JSON paths to check $exists

Enjoy!

**Comment by kerfab** [comment]
Update: it seems that my fix won't work exactly as MongoDB behavior for `$exists`

In fact, MongoDB is utterly broken when it comes to handling brackets notation with `$exists`: JSON paths which are valid for `$exists` won't work with MongoDB.

With the following document in a collection:

`{ [ [ {'c' : 1} ] ] }`   

What seems to work fine for MongoDB (tested on 3.4) is the purely dotted notation so that, eg:

`{'a[0][0].c' : {$exists:true}}` _will work in my patch_

 which does not work on MongoDB becomes:

`{'a.0.0.c' : {$exists:true}}` _requires an update in the patch to be handled properly_

which works on MongoDB.

What do you want to see implemented in mongoquery? I need to have mongoquery work fine on `$exists` so might update it accordingly.

Options are: handle both the brackets `[]` and dotted notations and ignore the fact that mongodb's results are inconsistent with said path, or ignore bracket notation (treated as part of key) and use dotted path notation only

Side note: I also noticed that my current patch does not recurse to find objects in arrays, I will make another upgrade on that one too.

**Comment by kapouille** [comment]
Hi,

Thanks for that detailed update. 

So far, you are the only user who has stretched the use of `$exists` to that extent so I'll let you steer the implementation (that's only fair as you're doing it).

I guess, for the sake of the other users, a disclaimer as to what differs from *MongoDB*'s implementation in the documentation blurb would be handy. 

Thanks for putting time into it!

**Comment by kerfab** [comment]
I will update the README file with my next update to include statements on the `$exists` operator. 

Do not take the pull request ATM, I will amend it with a more powerful locator which will be able to search within lists.

Thank you for the trust, I am sure I will come with a more powerful `$exists` implementation for `mongoquery`! I want `$exists` to be somehow as consistent as the mongodb version, so paths can be replayed in one/another :-)

**Comment by kapouille** [comment]
Thanks @kerfab , there's a new release , 1.12 with your changes in!

**Comment by kerfab** [comment]
Well, sorry for not coming back earlier, you updated just before my latest push ;)

I improved `$exists` and I fixed another issue:

```
>>> from mongoquery import Query
>>> q = Query({'a':{'$gt':3}})
>>> q.match({'a':'3'})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.4/dist-packages/mongoquery/__init__.py", line 27, in match
    return self._match(self._definition, entry)
  File "/usr/local/lib/python3.4/dist-packages/mongoquery/__init__.py", line 33, in _match
    for sub_operator, sub_condition in condition.items()
  File "/usr/local/lib/python3.4/dist-packages/mongoquery/__init__.py", line 33, in <genexpr>
    for sub_operator, sub_condition in condition.items()
  File "/usr/local/lib/python3.4/dist-packages/mongoquery/__init__.py", line 76, in _process_condition
    return self._match(condition, extracted_data)
  File "/usr/local/lib/python3.4/dist-packages/mongoquery/__init__.py", line 33, in _match
    for sub_operator, sub_condition in condition.items()
  File "/usr/local/lib/python3.4/dist-packages/mongoquery/__init__.py", line 33, in <genexpr>
    for sub_operator, sub_condition in condition.items()
  File "/usr/local/lib/python3.4/dist-packages/mongoquery/__init__.py", line 66, in _process_condition
    return getattr(self, "_" + operator[1:])(condition, entry)
  File "/usr/local/lib/python3.4/dist-packages/mongoquery/__init__.py", line 93, in _gt
    return entry > condition
TypeError: unorderable types: str() > int()
```

Now fixed with another Exception handle case in main code.

**Comment by kapouille** [comment]
Aha, right, sorry, I thought you had everything covered!

**Comment by kerfab** [comment]
I had it covered, but incompletely: so, the latest patch makes `$exists` work on nested arrays as well and fix two other bugs which will improve mongoquery's resilience with awkwards queries like out of ranges or mismatching data types.

Sorry for the delay, I integrated mongoquery in another library and so was busy finishing the core components of it before moving back on improving mongoquery ;-)

**Comment by kerfab** [comment]
Update: the PR is ready for integration, I don't think I will make any more commits now unless a new bug is spotted.

For info, here is the project I am working on: https://github.com/kerfab/jdic/

Mongoquery is used as main processor for the match*() functions in both mongo and jsonpath_ng drivers. 

It is WIP and is not ready for pip release yet, I am working on the pylint warnings now... :-) Jdic aims to be the cornerstone of a larger project for creating Business Documents over JSON.

**Comment by ch3pjw** [comment]
See #11

**Comment by kerfab** [comment]
Fixed in 1.3.0
